From b0b7c18f9dec0cb51513d69c83986ff2dcd384bc Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Fri, 17 Jul 2026 11:14:03 +0200
Subject: [PATCH] tests: shell: check for features not available in 5.4

JIRA: https://issues.redhat.com/browse/RHEL-190549
Upstream Status: nftables commit 883f79de486604e310c5093cfbe4d704794e4825

commit 883f79de486604e310c5093cfbe4d704794e4825
Author: Pablo Neira Ayuso <pablo@netfilter.org>
Date:   Thu May 29 11:49:18 2025 +0200

    tests: shell: check for features not available in 5.4

    5.4 -stable kernels report failures in these tests, this kernel version
    is lacking these feature.

    The bitshift requirement is needed by this ruleset:

     table ip x {
            set s13 {
                    typeof tcp option mptcp subtype
                    elements = { mp-join, dss }
            }

            chain y {
                    tcp option mptcp subtype @s13 accept
            }
     }

    which uses bitshift in its bytecode.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Phil Sutter <psutter@redhat.com>
---
 tests/shell/testcases/json/single_flag   | 1 +
 tests/shell/testcases/sets/elem_limit_0  | 2 +-
 tests/shell/testcases/sets/set_stmt      | 2 ++
 tests/shell/testcases/sets/typeof_sets_0 | 1 +
 4 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/tests/shell/testcases/json/single_flag b/tests/shell/testcases/json/single_flag
index 41fab63..43ae452 100755
--- a/tests/shell/testcases/json/single_flag
+++ b/tests/shell/testcases/json/single_flag
@@ -7,6 +7,7 @@
 #   recognized in input (checked against standard syntax input/output)
 
 # NFT_TEST_REQUIRES(NFT_TEST_HAVE_json)
+# NFT_TEST_REQUIRES(NFT_TEST_HAVE_table_flag_persist)
 
 set -e
 
diff --git a/tests/shell/testcases/sets/elem_limit_0 b/tests/shell/testcases/sets/elem_limit_0
index b57f927..ed60091 100755
--- a/tests/shell/testcases/sets/elem_limit_0
+++ b/tests/shell/testcases/sets/elem_limit_0
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-## requires EXPR
+# NFT_TEST_REQUIRES(NFT_TEST_HAVE_set_expr)
 
 set -e
 
diff --git a/tests/shell/testcases/sets/set_stmt b/tests/shell/testcases/sets/set_stmt
index 0433b67..ea50525 100755
--- a/tests/shell/testcases/sets/set_stmt
+++ b/tests/shell/testcases/sets/set_stmt
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+# NFT_TEST_REQUIRES(NFT_TEST_HAVE_set_expr)
+
 test_set_stmt() {
 	local i=$1
 	local stmt1=$2
diff --git a/tests/shell/testcases/sets/typeof_sets_0 b/tests/shell/testcases/sets/typeof_sets_0
index 3a4ed94..baf7c29 100755
--- a/tests/shell/testcases/sets/typeof_sets_0
+++ b/tests/shell/testcases/sets/typeof_sets_0
@@ -5,6 +5,7 @@
 # ways for declaration.
 
 # NFT_TEST_REQUIRES(NFT_TEST_HAVE_ip_options)
+# NFT_TEST_REQUIRES(NFT_TEST_HAVE_bitshift)
 
 set -e
 
